File: hangul.py
class DKSTHangul:
    def __init__(self):
        self._cho = 0
        self._jung = 0
        self._jong = 0
        self._buffer = []
        self._completed = []
        self.moa_jjiki_enabled = True

    def reset(self):
        self._cho = 0
        self._jung = 0
        self._jong = 0
        self._buffer = []
        self._completed = []

    def composed_string(self):
        if self._cho == 0 and self._jung == 0 and self._jong == 0:
            return ""
        return chr(self.current_syllable())

    def commit_string(self):
        res = "".join(self._completed)
        self._completed = []
        return res

    # Check types
    def is_cho(self, c): return 0x1100 <= c <= 0x1112
    def is_jung(self, c): return 0x1161 <= c <= 0x1175
    def is_jong(self, c): return 0x11A8 <= c <= 0x11C2

    def map_from_char(self, c):
        mapping = {
            'q': 0x1107, 'Q': 0x1108, # ㅂ, ㅃ
            'w': 0x110c, 'W': 0x110d, # ㅈ, ㅉ
            'e': 0x1103, 'E': 0x1104, # ㄷ, ㄸ
            'r': 0x1100, 'R': 0x1101, # ㄱ, ㄲ
            't': 0x1109, 'T': 0x110a, # ㅅ, ㅆ
            'y': 0x116d, 'Y': 0x116d, # ㅛ
            'u': 0x1167, 'U': 0x1167, # ㅕ
            'i': 0x1163, 'I': 0x1163, # ㅑ
            'o': 0x1162, 'O': 0x1164, # o=ㅐ, O=ㅒ
            'p': 0x1166, 'P': 0x1168, # p=ㅔ, P=ㅖ
            
            'a': 0x1106, 'A': 0x1106, # ㅁ
            's': 0x1102, 'S': 0x1102, # ㄴ
            'd': 0x110b, 'D': 0x110b, # ㅇ
            'f': 0x1105, 'F': 0x1105, # ㄹ
            'g': 0x1112, 'G': 0x1112, # ㅎ
            'h': 0x1169, 'H': 0x1169, # ㅗ
            'j': 0x1165, 'J': 0x1165, # ㅓ
            'k': 0x1161, 'K': 0x1161, # ㅏ
            'l': 0x1175, 'L': 0x1175, # ㅣ
            
            'z': 0x110f, 'Z': 0x110f, # ㅋ
            'x': 0x1110, 'X': 0x1110, # ㅌ
            'c': 0x110e, 'C': 0x110e, # ㅊ
            'v': 0x1111, 'V': 0x1111, # ㅍ
            'b': 0x1172, 'B': 0x1172, # ㅠ
            'n': 0x116e, 'N': 0x116e, # ㅜ
            'm': 0x1173, 'M': 0x1173, # ㅡ
        }
        return mapping.get(c, 0)

    def current_syllable(self):
        if self._cho == 0 and self._jung == 0 and self._jong == 0:
            return 0
        
        # Independent Jamo
        if self._cho and not self._jung and not self._jong:
            return self.compatibility_jamo(self._cho)
        if not self._cho and self._jung and not self._jong:
            return self.compatibility_jamo(self._jung)
        
        cho_idx = self.cho_index(self._cho) if self._cho else -1
        jung_idx = self.jung_index(self._jung) if self._jung else -1
        jong_idx = self.jong_index(self._jong) if self._jong else 0
        
        if cho_idx != -1 and jung_idx != -1:
            return 0xAC00 + (cho_idx * 21 * 28) + (jung_idx * 28) + jong_idx
        
        # Jung Only (Moa-jjiki transitional)
        if jung_idx != -1 and cho_idx == -1:
            return self.compatibility_jamo(self._jung)
            
        return 0

    def compatibility_jamo(self, u):
        if 0x1100 <= u <= 0x1112:
            base = u - 0x1100
            # ㄱ ㄲ ㄴ ㄷ ㄸ ㄹ ㅁ ㅂ ㅃ ㅅ ㅆ ㅇ ㅈ ㅉ ㅊ ㅋ ㅌ ㅍ ㅎ
            # 3131 3132 3134 3137 3138 3139 3141 3142 3143 3145 3146 3147 3148 3149 314A 314B 314C 314D 314E
            map_arr = [
                0x3131, 0x3132, 0x3134, 0x3137, 0x3138, 0x3139, 0x3141, 0x3142, 0x3143,
                0x3145, 0x3146, 0x3147, 0x3148, 0x3149, 0x314A, 0x314B, 0x314C, 0x314D, 0x314E
            ]
            if base < len(map_arr): return map_arr[base]
            
        if 0x1161 <= u <= 0x1175:
            base = u - 0x1161
            # ㅏ ㅐ ㅑ ㅒ ㅓ ㅔ ㅕ ㅖ ㅗ ㅘ ㅙ ㅚ ㅛ ㅜ ㅝ ㅞ ㅟ ㅠ ㅡ ㅢ ㅣ
            # 314F 3150 3151 3152 3153 3154 3155 3156 3157 3158 3159 315A 315B 315C 315D 315E 315F 3160 3161 3162 3163
            map_arr = [
                0x314F, 0x3150, 0x3151, 0x3152, 0x3153, 0x3154, 0x3155, 0x3156, 0x3157,
                0x3158, 0x3159, 0x315A, 0x315B, 0x315C, 0x315D, 0x315E, 0x315F, 0x3160,
                0x3161, 0x3162, 0x3163
            ]
            if base < len(map_arr): return map_arr[base]
            
        return u

    def cho_index(self, c):
        if 0x1100 <= c <= 0x1112: return c - 0x1100
        return -1

    def jung_index(self, c):
        if 0x1161 <= c <= 0x1175: return c - 0x1161
        return -1

    def jong_index(self, c):
        if 0x11A8 <= c <= 0x11C2: return c - 0x11A8 + 1
        return 0

    def backspace(self):
        if self._cho == 0 and self._jung == 0 and self._jong == 0:
            return False
        
        if self._jong != 0:
            j1, j2 = self.split_jong(self._jong)
            if j2 != 0:
                self._jong = j1
            else:
                self._jong = 0
            return True
            
        if self._jung != 0:
            j1, j2 = self.split_jung(self._jung)
            if j2 != 0:
                self._jung = j1
            else:
                self._jung = 0
            return True
            
        if self._cho != 0:
            self._cho = 0
            return True
            
        return False

    def process_code(self, key_char):
        # key_char should be a character 'a', 'b', etc.
        hangul = self.map_from_char(key_char)
        
        if hangul == 0:
            # Commit current if exists
            if self._cho or self._jung or self._jong:
                self._completed.append(chr(self.current_syllable()))
                self._cho = 0; self._jung = 0; self._jong = 0
            return False
            
        if self.is_cho(hangul):
            if self._jung == 0:
                if self._cho == 0:
                    self._cho = hangul
                else:
                    self._completed.append(chr(self.compatibility_jamo(self._cho)))
                    self._cho = hangul
            else:
                if self._jong == 0:
                    if self._cho == 0:
                        if self.moa_jjiki_enabled:
                            self._cho = hangul
                            return True
                        else:
                            self._completed.append(chr(self.current_syllable()))
                            self._cho = hangul; self._jung = 0; self._jong = 0
                            return True
                    
                    as_jong = self.cho_to_jong(hangul)
                    if as_jong:
                        self._jong = as_jong
                    else:
                        self._completed.append(chr(self.current_syllable()))
                        self._cho = hangul; self._jung = 0; self._jong = 0
                else:
                    # Cho+Jung+Jong
                    compound = self.combine_jong(self._jong, self.cho_to_jong(hangul))
                    if compound:
                        self._jong = compound
                    else:
                        self._completed.append(chr(self.current_syllable()))
                        self._cho = hangul; self._jung = 0; self._jong = 0
                        
        elif self.is_jung(hangul):
            if self._jong:
                j1, j2 = self.split_jong(self._jong)
                if j2:
                    self._jong = j1
                    next_cho = self.jong_to_cho(j2)
                    self._completed.append(chr(self.current_syllable()))
                    self._cho = next_cho; self._jung = hangul; self._jong = 0
                else:
                    next_cho = self.jong_to_cho(self._jong)
                    self._jong = 0
                    self._completed.append(chr(self.current_syllable()))
                    self._cho = next_cho; self._jung = hangul; self._jong = 0
                    
            elif self._jung:
                compound = self.combine_jung(self._jung, hangul)
                if compound:
                    self._jung = compound
                else:
                    self._completed.append(chr(self.current_syllable()))
                    self._cho = 0; self._jung = hangul; self._jong = 0
            else:
                self._jung = hangul # Independent or with Cho
                
        return True

    def cho_to_jong(self, c):
        mapping = {
            0x1100: 0x11A8, 0x1101: 0x11A9, 0x1102: 0x11AB, 0x1103: 0x11AE,
            0x1105: 0x11AF, 0x1106: 0x11B7, 0x1107: 0x11B8, 0x1109: 0x11BA,
            0x110A: 0x11BB, 0x110B: 0x11BC, 0x110C: 0x11BD, 0x110E: 0x11BE,
            0x110F: 0x11BF, 0x1110: 0x11C0, 0x1111: 0x11C1, 0x1112: 0x11C2
        }
        return mapping.get(c, 0)

    def jong_to_cho(self, c):
        mapping = {
            0x11A8: 0x1100, 0x11A9: 0x1101, 0x11AB: 0x1102, 0x11AE: 0x1103,
            0x11AF: 0x1105, 0x11B7: 0x1106, 0x11B8: 0x1107, 0x11BA: 0x1109,
            0x11BB: 0x110A, 0x11BC: 0x110B, 0x11BD: 0x110C, 0x11BE: 0x110E,
            0x11BF: 0x110F, 0x11C0: 0x1110, 0x11C1: 0x1111, 0x11C2: 0x1112
        }
        return mapping.get(c, 0)
        
    def combine_jung(self, a, b):
        if a == 0x1169 and b == 0x1161: return 0x116A # ㅘ
        if a == 0x1169 and b == 0x1162: return 0x116B # ㅙ
        if a == 0x1169 and b == 0x1175: return 0x116C # ㅚ
        if a == 0x116e and b == 0x1165: return 0x116F # ㅝ
        if a == 0x116e and b == 0x1166: return 0x1170 # ㅞ
        if a == 0x116e and b == 0x1175: return 0x1171 # ㅟ
        if a == 0x1173 and b == 0x1175: return 0x1174 # ㅢ
        return 0

    def split_jung(self, c):
        mapping = {
            0x116A: (0x1169, 0x1161), 0x116B: (0x1169, 0x1162), 0x116C: (0x1169, 0x1175),
            0x116F: (0x116e, 0x1165), 0x1170: (0x116e, 0x1166), 0x1171: (0x116e, 0x1175),
            0x1174: (0x1173, 0x1175)
        }
        return mapping.get(c, (c, 0))

    def combine_jong(self, a, b):
        if a == 0x11A8 and b == 0x11BA: return 0x11AA # ㄳ
        if a == 0x11AB and b == 0x11BD: return 0x11AC # ㄵ
        if a == 0x11AB and b == 0x11C2: return 0x11AD # ㄶ
        if a == 0x11AF and b == 0x11A8: return 0x11B0 # ㄺ
        if a == 0x11AF and b == 0x11B7: return 0x11B1 # ㄻ
        if a == 0x11AF and b == 0x11B8: return 0x11B2 # ㄼ
        if a == 0x11AF and b == 0x11BA: return 0x11B3 # ㄽ
        if a == 0x11AF and b == 0x11C0: return 0x11B4 # ㄾ
        if a == 0x11AF and b == 0x11C1: return 0x11B5 # ㄿ
        if a == 0x11AF and b == 0x11C2: return 0x11B6 # ㅀ
        if a == 0x11B8 and b == 0x11BA: return 0x11B9 # ㅄ
        return 0

    def split_jong(self, c):
        mapping = {
            0x11AA: (0x11A8, 0x11BA), 0x11AC: (0x11AB, 0x11BD), 0x11AD: (0x11AB, 0x11C2),
            0x11B0: (0x11AF, 0x11A8), 0x11B1: (0x11AF, 0x11B7), 0x11B2: (0x11AF, 0x11B8),
            0x11B3: (0x11AF, 0x11BA), 0x11B4: (0x11AF, 0x11C0), 0x11B5: (0x11AF, 0x11C1),
            0x11B6: (0x11AF, 0x11C2), 0x11B9: (0x11B8, 0x11BA)
        }
        return mapping.get(c, (c, 0))

File: hangul.c

#include "hangul.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Jamo Ranges
#define IS_CHO(c) (0x1100 <= (c) && (c) <= 0x1112)
#define IS_JUNG(c) (0x1161 <= (c) && (c) <= 0x1175)
#define IS_JONG(c) (0x11A8 <= (c) && (c) <= 0x11C2)

void dkst_hangul_init(DKSTHangul *h) {
  h->cho = 0;
  h->jung = 0;
  h->jong = 0;
  h->buffer = g_string_new("");
  h->completed = g_string_new("");
  h->moa_jjiki_enabled = true;
  h->backspace_mode = DKST_BACKSPACE_JASO;
}

void dkst_hangul_reset(DKSTHangul *h) {
  h->cho = 0;
  h->jung = 0;
  h->jong = 0;
  // buffer usually tracks raw input history, but logic here is state-based.
  // We clear it for consistency.
  g_string_truncate(h->buffer, 0);
  // Do NOT clear completed here automatically? logic says commit consumes it.
}

void dkst_hangul_free(DKSTHangul *h) {
  if (h->buffer)
    g_string_free(h->buffer, TRUE);
  if (h->completed)
    g_string_free(h->completed, TRUE);
}

// Map char to Jamo
static uint32_t map_key(char c) {
  switch (c) {
  case 'q':
    return 0x1107;
  case 'Q':
    return 0x1108; // ㅂ, ㅃ
  case 'w':
    return 0x110c;
  case 'W':
    return 0x110d; // ㅈ, ㅉ
  case 'e':
    return 0x1103;
  case 'E':
    return 0x1104; // ㄷ, ㄸ
  case 'r':
    return 0x1100;
  case 'R':
    return 0x1101; // ㄱ, ㄲ
  case 't':
    return 0x1109;
  case 'T':
    return 0x110a; // ㅅ, ㅆ
  case 'y':
    return 0x116d;
  case 'Y':
    return 0x116d; // ㅛ
  case 'u':
    return 0x1167;
  case 'U':
    return 0x1167; // ㅕ
  case 'i':
    return 0x1163;
  case 'I':
    return 0x1163; // ㅑ
  case 'o':
    return 0x1162;
  case 'O':
    return 0x1164; // ㅐ, ㅒ
  case 'p':
    return 0x1166;
  case 'P':
    return 0x1168; // ㅔ, ㅖ

  case 'a':
    return 0x1106;
  case 'A':
    return 0x1106; // ㅁ
  case 's':
    return 0x1102;
  case 'S':
    return 0x1102; // ㄴ
  case 'd':
    return 0x110b;
  case 'D':
    return 0x110b; // ㅇ
  case 'f':
    return 0x1105;
  case 'F':
    return 0x1105; // ㄹ
  case 'g':
    return 0x1112;
  case 'G':
    return 0x1112; // ㅎ
  case 'h':
    return 0x1169;
  case 'H':
    return 0x1169; // ㅗ
  case 'j':
    return 0x1165;
  case 'J':
    return 0x1165; // ㅓ
  case 'k':
    return 0x1161;
  case 'K':
    return 0x1161; // ㅏ
  case 'l':
    return 0x1175;
  case 'L':
    return 0x1175; // ㅣ

  case 'z':
    return 0x110f;
  case 'Z':
    return 0x110f; // ㅋ
  case 'x':
    return 0x1110;
  case 'X':
    return 0x1110; // ㅌ
  case 'c':
    return 0x110e;
  case 'C':
    return 0x110e; // ㅊ
  case 'v':
    return 0x1111;
  case 'V':
    return 0x1111; // ㅍ
  case 'b':
    return 0x1172;
  case 'B':
    return 0x1172; // ㅠ
  case 'n':
    return 0x116e;
  case 'N':
    return 0x116e; // ㅜ
  case 'm':
    return 0x1173;
  case 'M':
    return 0x1173; // ㅡ
  default:
    return 0;
  }
}

static uint32_t compatibility_jamo(uint32_t u) {
  if (0x1100 <= u && u <= 0x1112) {
    // Simple offset mapping for Chosung to Compatibility Jamo
    static const uint32_t map[] = {0x3131, 0x3132, 0x3134, 0x3137, 0x3138,
                                   0x3139, 0x3141, 0x3142, 0x3143, 0x3145,
                                   0x3146, 0x3147, 0x3148, 0x3149, 0x314A,
                                   0x314B, 0x314C, 0x314D, 0x314E};
    int idx = u - 0x1100;
    if (idx >= 0 && idx < 19)
      return map[idx];
  }
  if (0x1161 <= u && u <= 0x1175) {
    static const uint32_t map[] = {
        0x314F, 0x3150, 0x3151, 0x3152, 0x3153, 0x3154, 0x3155,
        0x3156, 0x3157, 0x3158, 0x3159, 0x315A, 0x315B, 0x315C,
        0x315D, 0x315E, 0x315F, 0x3160, 0x3161, 0x3162, 0x3163};
    int idx = u - 0x1161;
    if (idx >= 0 && idx < 21)
      return map[idx];
  }
  return u;
}

static int cho_index(uint32_t c) {
  if (0x1100 <= c && c <= 0x1112)
    return c - 0x1100;
  return -1;
}
static int jung_index(uint32_t c) {
  if (0x1161 <= c && c <= 0x1175)
    return c - 0x1161;
  return -1;
}
static int jong_index(uint32_t c) {
  if (0x11A8 <= c && c <= 0x11C2)
    return c - 0x11A8 + 1;
  return 0; // 0 means no jongseong
}

static uint32_t cho_to_jong(uint32_t c) {
  // Mapping table from py
  switch (c) {
  case 0x1100:
    return 0x11A8;
  case 0x1101:
    return 0x11A9;
  case 0x1102:
    return 0x11AB;
  case 0x1103:
    return 0x11AE;
  case 0x1105:
    return 0x11AF;
  case 0x1106:
    return 0x11B7;
  case 0x1107:
    return 0x11B8;
  case 0x1109:
    return 0x11BA;
  case 0x110A:
    return 0x11BB;
  case 0x110B:
    return 0x11BC;
  case 0x110C:
    return 0x11BD;
  case 0x110E:
    return 0x11BE;
  case 0x110F:
    return 0x11BF;
  case 0x1110:
    return 0x11C0;
  case 0x1111:
    return 0x11C1;
  case 0x1112:
    return 0x11C2;
  default:
    return 0;
  }
}
static uint32_t jong_to_cho(uint32_t c) {
  switch (c) {
  case 0x11A8:
    return 0x1100;
  case 0x11A9:
    return 0x1101;
  case 0x11AB:
    return 0x1102;
  case 0x11AE:
    return 0x1103;
  case 0x11AF:
    return 0x1105;
  case 0x11B7:
    return 0x1106;
  case 0x11B8:
    return 0x1107;
  case 0x11BA:
    return 0x1109;
  case 0x11BB:
    return 0x110A;
  case 0x11BC:
    return 0x110B;
  case 0x11BD:
    return 0x110C;
  case 0x11BE:
    return 0x110E;
  case 0x11BF:
    return 0x110F;
  case 0x11C0:
    return 0x1110;
  case 0x11C1:
    return 0x1111;
  case 0x11C2:
    return 0x1112;
  default:
    return 0;
  }
}

static uint32_t combine_jung(uint32_t a, uint32_t b) {
  if (a == 0x1169 && b == 0x1161)
    return 0x116A; // ㅘ
  if (a == 0x1169 && b == 0x1162)
    return 0x116B; // ㅙ
  if (a == 0x1169 && b == 0x1175)
    return 0x116C; // ㅚ
  if (a == 0x116e && b == 0x1165)
    return 0x116F; // ㅝ
  if (a == 0x116e && b == 0x1166)
    return 0x1170; // ㅞ
  if (a == 0x116e && b == 0x1175)
    return 0x1171; // ㅟ
  if (a == 0x1173 && b == 0x1175)
    return 0x1174; // ㅢ
  return 0;
}

static void split_jung(uint32_t c, uint32_t *j1, uint32_t *j2) {
  *j1 = c;
  *j2 = 0;
  if (c == 0x116A) {
    *j1 = 0x1169;
    *j2 = 0x1161;
  } else if (c == 0x116B) {
    *j1 = 0x1169;
    *j2 = 0x1162;
  } else if (c == 0x116C) {
    *j1 = 0x1169;
    *j2 = 0x1175;
  } else if (c == 0x116F) {
    *j1 = 0x116e;
    *j2 = 0x1165;
  } else if (c == 0x1170) {
    *j1 = 0x116e;
    *j2 = 0x1166;
  } else if (c == 0x1171) {
    *j1 = 0x116e;
    *j2 = 0x1175;
  } else if (c == 0x1174) {
    *j1 = 0x1173;
    *j2 = 0x1175;
  }
}

static uint32_t combine_jong(uint32_t a, uint32_t b) {
  if (a == 0x11A8 && b == 0x11BA)
    return 0x11AA; // ㄳ
  if (a == 0x11AB && b == 0x11BD)
    return 0x11AC; // ㄵ
  if (a == 0x11AB && b == 0x11C2)
    return 0x11AD; // ㄶ
  if (a == 0x11AF && b == 0x11A8)
    return 0x11B0; // ㄺ
  if (a == 0x11AF && b == 0x11B7)
    return 0x11B1; // ㄻ
  if (a == 0x11AF && b == 0x11B8)
    return 0x11B2; // ㄼ
  if (a == 0x11AF && b == 0x11BA)
    return 0x11B3; // ㄽ
  if (a == 0x11AF && b == 0x11C0)
    return 0x11B4; // ㄾ
  if (a == 0x11AF && b == 0x11C1)
    return 0x11B5; // ㄿ
  if (a == 0x11AF && b == 0x11C2)
    return 0x11B6; // ㅀ
  if (a == 0x11B8 && b == 0x11BA)
    return 0x11B9; // ㅄ
  return 0;
}

static void split_jong(uint32_t c, uint32_t *j1, uint32_t *j2) {
  *j1 = c;
  *j2 = 0;
  if (c == 0x11AA) {
    *j1 = 0x11A8;
    *j2 = 0x11BA;
  } else if (c == 0x11AC) {
    *j1 = 0x11AB;
    *j2 = 0x11BD;
  } else if (c == 0x11AD) {
    *j1 = 0x11AB;
    *j2 = 0x11C2;
  } else if (c == 0x11B0) {
    *j1 = 0x11AF;
    *j2 = 0x11A8;
  } else if (c == 0x11B1) {
    *j1 = 0x11AF;
    *j2 = 0x11B7;
  } else if (c == 0x11B2) {
    *j1 = 0x11AF;
    *j2 = 0x11B8;
  } else if (c == 0x11B3) {
    *j1 = 0x11AF;
    *j2 = 0x11BA;
  } else if (c == 0x11B4) {
    *j1 = 0x11AF;
    *j2 = 0x11C0;
  } else if (c == 0x11B5) {
    *j1 = 0x11AF;
    *j2 = 0x11C1;
  } else if (c == 0x11B6) {
    *j1 = 0x11AF;
    *j2 = 0x11C2;
  } else if (c == 0x11B9) {
    *j1 = 0x11B8;
    *j2 = 0x11BA;
  }
}

uint32_t dkst_hangul_current_syllable(DKSTHangul *h) {
  if (h->cho == 0 && h->jung == 0 && h->jong == 0)
    return 0;

  // Independent Jamo
  if (h->cho && !h->jung && !h->jong)
    return compatibility_jamo(h->cho);
  if (!h->cho && h->jung && !h->jong)
    return compatibility_jamo(h->jung);

  int c = (h->cho) ? cho_index(h->cho) : -1;
  int j = (h->jung) ? jung_index(h->jung) : -1;
  int k = (h->jong) ? jong_index(h->jong) : 0;

  if (c != -1 && j != -1) {
    return 0xAC00 + (c * 21 * 28) + (j * 28) + k;
  }

  // Moa-jjik-gi partial
  if (j != -1 && c == -1)
    return compatibility_jamo(h->jung);

  return 0; // Should not happen with valid logic
}

bool dkst_hangul_backspace(DKSTHangul *h) {
  if (h->cho == 0 && h->jung == 0 && h->jong == 0)
    return false;

  // Character Mode: Wipe everything
  if (h->backspace_mode == DKST_BACKSPACE_CHAR) {
    h->cho = 0;
    h->jung = 0;
    h->jong = 0;
    return true;
  }

  // Jaso Mode: Detailed breakdown
  if (h->jong != 0) {
    uint32_t j1, j2;
    split_jong(h->jong, &j1, &j2);
    if (j2)
      h->jong = j1;
    else
      h->jong = 0;
    return true;
  }

  if (h->jung != 0) {
    uint32_t j1, j2;
    split_jung(h->jung, &j1, &j2);
    if (j2)
      h->jung = j1;
    else
      h->jung = 0;
    return true;
  }

  if (h->cho != 0) {
    h->cho = 0;
    return true;
  }
  return false;
}

// Append unicode char to buffer string
static void append_unichar(GString *s, uint32_t u) {
  if (u == 0)
    return;
  g_string_append_unichar(s, u);
}

bool dkst_hangul_process(DKSTHangul *h, char key) {
  uint32_t hangul = map_key(key);

  if (hangul == 0) {
    // Not a hangul key. Commit current and return false (not consumed)
    if (h->cho || h->jung || h->jong) {
      uint32_t syl = dkst_hangul_current_syllable(h);
      append_unichar(h->completed, syl);
      dkst_hangul_reset(h);
    }
    return false;
  }

  if (IS_CHO(hangul)) {
    if (h->jung == 0) {
      if (h->cho == 0) {
        h->cho = hangul;
      } else {
        append_unichar(h->completed, compatibility_jamo(h->cho));
        h->cho = hangul;
        // No need to reset others as they are 0
      }
    } else {
      // We have Cho and Jung
      if (h->jong == 0) {
        if (h->cho == 0) {
          // Jung only present.
          if (h->moa_jjiki_enabled) {
            h->cho = hangul;
            return true;
          } else {
            append_unichar(h->completed, dkst_hangul_current_syllable(h));
            dkst_hangul_reset(h);
            h->cho = hangul;
            return true;
          }
        }

        // Standard case: Cho+Jung. Incoming Cho might be Jong.
        uint32_t as_jong = cho_to_jong(hangul);
        if (as_jong) {
          h->jong = as_jong;
        } else {
          append_unichar(h->completed, dkst_hangul_current_syllable(h));
          dkst_hangul_reset(h);
          h->cho = hangul;
        }
      } else {
        // Cho+Jung+Jong. Incoming Cho might combine with Jong.
        uint32_t compound = combine_jong(h->jong, cho_to_jong(hangul));
        if (compound) {
          h->jong = compound;
        } else {
          append_unichar(h->completed, dkst_hangul_current_syllable(h));
          dkst_hangul_reset(h);
          h->cho = hangul;
        }
      }
    }
  } else if (IS_JUNG(hangul)) {
    if (h->jong) {
      uint32_t j1, j2;
      split_jong(h->jong, &j1, &j2);
      if (j2) {
        // Complex jong. Split it. current becomes Cho+Jung+J1. Next is J2(as
        // Cho)+NewJung
        h->jong = j1;
        uint32_t syl = dkst_hangul_current_syllable(h);
        append_unichar(h->completed, syl);

        uint32_t next_cho = jong_to_cho(j2);
        dkst_hangul_reset(h);
        h->cho = next_cho;
        h->jung = hangul;
      } else {
        // Simple jong. It moves to next char as Cho.
        uint32_t next_cho = jong_to_cho(h->jong);
        h->jong = 0;
        uint32_t syl = dkst_hangul_current_syllable(h);
        append_unichar(h->completed, syl);

        dkst_hangul_reset(h); // Clear
        h->cho = next_cho;
        h->jung = hangul;
      }
    } else if (h->jung) {
      uint32_t compound = combine_jung(h->jung, hangul);
      if (compound) {
        h->jung = compound;
      } else {
        append_unichar(h->completed, dkst_hangul_current_syllable(h));
        dkst_hangul_reset(h);
        h->jung = hangul; // Assuming independent jung valid or moa-jjiki start
      }
    } else {
      // Cho might be set or not
      h->jung = hangul;
    }
  }

  return true;
}

char *dkst_hangul_get_commit_string(DKSTHangul *h) {
  if (h->completed->len == 0)
    return NULL;
  char *ret = g_strdup(h->completed->str);
  g_string_truncate(h->completed, 0);
  return ret;
}

bool dkst_hangul_has_composed(DKSTHangul *h) {
  return (h->cho || h->jung || h->jong);
}
